### PR TITLE
Cargo bump, Instruments selection, and refactor

### PR DESCRIPTION
**Pull request by bosr**
Hello Colin,

I love your project, it really opened my eyes on what's possible with Instruments, even on Rust!

I wanted to provide a few improvements, but then I ended up refactoring the code much more than I had expected!

I don't think looking at the diff can provide a good overview of what has changed because I have moved too many pieces around. I think it's much easier to clone and read the code to see the logic, the names, etc.

In a nutshell
- variable names have changed from `opts`, `args` to `cargo_options`, `app_config`, etc, because I was having a hard time following
- more comments
- changed a bit the options
- Xcode Instruments selection mechanism (`xctrace` or `instruments`)
- gather templates into a catalog `struct`
- no default template: provide interactive selection if no template is provided

I have not changed much to the cargo manipulations, I just updated to 0.52's `UnitOutput` and leveraged `workspace.root()` to get the path.

I really hope you will like it, discuss and merge, but, truly, I'd understand such a refactoring looks inopportune and I apologize if that's the case.

Best regards

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,50 +1,81 @@
...
-`cargo-instruments` is the glue between cargo and Xcode's bundled profiling
+`cargo-instruments` is the glue between Cargo and Xcode's bundled profiling
 suite. It allows you to easily profile any binary in your crate, generating
 files that can be viewed in the Instruments app.
 
 ![Instruments Time Profiler](https://raw.githubusercontent.com/cmyr/cargo-instruments/screenshots/instruments_time1.png)
 ![Instruments System Trace](https://raw.githubusercontent.com/cmyr/cargo-instruments/screenshots/instruments_sys1.png)
 
+This is a proposal refactoring of [Colin Rofls' excellent
+work](https://github.com/cmyr/cargo-instruments).
```
Comment (cmyr) [inline README.md:13]: We won't need this comment when we merge 😉

File: README.md [hunk 2]
```diff
@@ -1,50 +1,81 @@
...
-```bash
-brew install cargo-instruments
+This will install Colin's `0.3.4` published version.
+
+```sh
+$ brew install cargo-instruments
 ```
 
+To install this fork's version (`0.4.0`), please install from source.
+
```
Comment (cmyr) [inline README.md:47]: ditto

File: README.md [hunk 3]
```diff
@@ -60,28 +91,97 @@ can append the following section in your `Cargo.toml`.
...
 ```sh
-# profile examples/my_example.rs, with the default template,
+# profile examples/my_example.rs, with the Allocations template,
 # for 10 seconds, and open the trace when finished
-$ cargo instruments --example my_example --limit 10000 --open
+$ cargo instruments -t Allocations --example my_example --time-limit 10000 --open
+```
+
+```sh
+# omitting to provide a template will display a `dialoguer` menu
```
Comment (cmyr) [inline README.md:182]:
```suggestion
# omitting to provide a template will display a `dialoguer` menu
```

is 'dialogeur' a typo, or just a term I'm not familiar with?

File: src/app.rs [hunk 4]
```diff
@@ -11,101 +11,144 @@ use cargo::util::interning::InternedString;
...
-    use cargo::util::important_paths::find_root_manifest_for_wd;
-    instruments::check_existence()?;
-
-    if args.list {
-        let list = instruments::list()?;
-        println!("{}", list);
+use crate::opt::{AppConfig, CargoOpts, Target};
+
+/// Main entrance point, after args have been parsed.
+///
```
Comment (cmyr) [inline src/app.rs:17]:
```suggestion
```

File: src/app.rs [hunk 5]
```diff
@@ -11,101 +11,144 @@ use cargo::util::interning::InternedString;
...
+        println!("{}", instruments::render_template_catalog(&catalog));
         return Ok(());
     }
 
-    let cfg = Config::default()?;
-    let manifest_path = find_root_manifest_for_wd(cfg.cwd())?;
-    let workspace = Workspace::new(&manifest_path, &cfg)?;
-    let workspace_root = manifest_path.parent().unwrap().to_owned();
+    // 3. Build the specified target
+    use cargo::util::important_paths::find_root_manifest_for_wd;
```
Comment (cmyr) [inline src/app.rs:30]: I would have `use cargo::util::important_paths` at the top of the file, and then call `important_paths::find_root_manifest_for_wd` below.

File: src/app.rs [hunk 6]
```diff
@@ -118,29 +161,13 @@ fn make_compile_opts(cargo_args: &CargoOpts, cfg: &Config) -> Result<CompileOpti
...
-    } else {
-        Ok(())
-    }
+    Ok(compile_options)
 }
 
-fn open_file(file: &PathBuf) -> Result<()> {
-    let status = Command::new("open").arg(file).status()?;
+/// Launch Xcode Instruments on the provided trace file.
+///
```
Comment (cmyr) [inline src/app.rs:168]:
```suggestion
```

File: src/instruments.rs [hunk 7]
```diff
@@ -1,133 +1,331 @@
...
+use crate::opt::AppConfig;
 
-/// Check that `instruments` is in $PATH.
-pub(crate) fn check_existence() -> Result<()> {
-    let path = ["/", "usr", "bin", "instruments"].iter().collect::<PathBuf>();
-    if path.exists() {
-        Ok(())
-    } else {
+/// Holds available templates.
+///
```
Comment (cmyr) [inline src/instruments.rs:14]:
```suggestion
```

File: src/instruments.rs [hunk 8]
```diff
@@ -1,133 +1,331 @@
...
+}
+
+impl TemplateCatalog {
+    /// Return a vector of all templates.
+    fn all_templates(&self) -> Vec<&str> {
+        self.standard_templates
+            .iter()
+            .chain(self.custom_templates.iter())
+            .map(|value| value.as_str())
+            .collect::<Vec<_>>()
```
Comment (cmyr) [inline src/instruments.rs:27]: this really doesn't matter in this particular case, but as a general observation I would prefer the pattern of returning `impl Iterator<Item=&str>` here, and then if the caller wants a vec they can handle the collecting, but if they didn't need a vec then we let them skip an allocation.

Again this isn't important here, but thought I'd mention it.

File: src/instruments.rs [hunk 9]
```diff
@@ -1,133 +1,331 @@
...
+    }
+}
+
+/// Represents the Xcode Instrument version detected.
+pub enum XcodeInstruments {
```
Comment (cmyr) [inline src/instruments.rs:32]: Is this `pub` in a meaningful way? Since this is a `bin` crate I suppose it doesn't really matter? Otherwise I would suggest `pub(crate)`.

File: src/instruments.rs [hunk 10]
```diff
@@ -1,133 +1,331 @@
...
+
+impl XcodeInstruments {
+    /// Detects which version of Xcode Instruments is installed and if it can be launched.
+    pub(crate) fn detect() -> Result<XcodeInstruments> {
+        let cur_version = get_macos_version()?;
+        let macos_xctrace_version = Version::parse("10.15.0").unwrap();
+
+        if cur_version >= macos_xctrace_version {
+            // This is the check used by Homebrew,see
+            // https://github.com/Homebrew/install/blob/master/install.sh#L216
```
Comment (cmyr) [inline src/instruments.rs:45]: cool, I think this is good justification.

File: src/instruments.rs [hunk 11]
```diff
@@ -1,133 +1,331 @@
...
+            }
+        };
+
+        if !status.success() {
+            return Err(anyhow!(
+                "Could not list templates. Please check your Xcode Instruments installation."
+            ));
+        }
+
+        let templates_str = String::from_utf8(stdout)?;
```
Comment (cmyr) [inline src/instruments.rs:108]: another silly thing but I would do `std::str::from_utf8` and avoid the allocation, although it is going to be meaningless in this particular case. Good muscle memory, though?

File: src/instruments.rs [hunk 12]
```diff
@@ -1,133 +1,331 @@
...
-    let now = Local::now();
-    let fmt = "%FT%T";
-    now.format(&fmt)
+    Ok(trace_filepath)
 }
 
-fn resolve_template(args: &Opts) -> &str {
-    match args.template.as_str() {
+/// Return the complete template name, replacing abbreviation if provided.
+///
```
Comment (cmyr) [inline src/instruments.rs:326]:
```suggestion
```

File: src/instruments.rs [hunk 13]
```diff
@@ -136,12 +334,71 @@ fn resolve_template(args: &Opts) -> &str {
     }
 }
 
-fn abbrev_name(template: &str) -> Option<&'static str> {
-    match template {
+/// Return the template name abbreviation if available.
+///
```
Comment (cmyr) [inline src/instruments.rs:338]:
```suggestion
```

File: src/instruments.rs [hunk 14]
```diff
@@ -136,12 +334,71 @@ fn resolve_template(args: &Opts) -> &str {
...
         "Allocations" => Some("alloc"),
         "File Activity" => Some("io"),
         "System Trace" => Some("sys"),
         _ => None,
     }
 }
+
+/// Profile the target binary at `binary_filepath` and write results at
+/// `trace_filepath`.
+///
```
Comment (cmyr) [inline src/instruments.rs:351]:
```suggestion
```

File: README.md (hunk 1 above)
Comment (bosr) [reply README.md:13]: Absolutely

File: README.md (hunk 3 above)
Comment (bosr) [reply README.md:182]: dialoguer is a crate for interactive menus: https://crates.io/crates/dialoguer
I will improve the wording.

File: src/instruments.rs (hunk 8 above)
Comment (bosr) [reply src/instruments.rs:27]: Thanks, I also tend to favor composability and minimize allocations, but as you point out, here I think it's ok.

File: src/instruments.rs (hunk 9 above)
Comment (bosr) [reply src/instruments.rs:32]: No, no there was no meaningful reason. Thanks

File: src/instruments.rs (hunk 11 above)
Comment (bosr) [reply src/instruments.rs:108]: Thanks! I completely missed it.

File: src/instruments.rs (hunk 9 above)
Comment (bosr) [reply src/instruments.rs:32]: Eh, I replied too quickly, `XcodeInstruments` is used from `app.rs`.

File: README.md (hunk 3 above)
Comment (cmyr) [reply README.md:182]: ah cool I haven't actually tried to run this yet, I should do that before I merge 😉

File: src/instruments.rs (hunk 11 above)
Comment (bosr) [reply src/instruments.rs:108]: Just for completeness, `String::from_utf8` moves in the `stdout` vector of `u8`. According to the [docs](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8) this function will take care of not copying the vector. That's pretty cool! If you don't mind, I'll leave it as is.

**Comment by bosr** [comment]
@cmyr I think I addressed all your comments, thanks!!
I also reverted the Cargo.toml's version and authors, so you decide the version.

Best

**Comment by cmyr** [comment]
Cool, will give this a closer look when I have a chance, either tonight or later this week. 👌

**Comment by bosr** [comment]
I just pushed a couple of changes, in particular using `join` on PathBuf and Path, which makes path manipulation more concise and clear.

**Comment by cmyr** [comment]
@bosr feel free to include the original changes to the version and author fields, I think both are appropriate, and I'll want to do a new release shortly after merging this.

File: src/instruments.rs [hunk 15]
```diff
@@ -1,133 +1,314 @@
...
+        }
+
+        let custom_templates = templates_iter
+            .map(|line| line.trim())
+            .skip_while(|line| line.starts_with('=') || line.is_empty())
+            .map(|line| line.into())
+            .collect::<Vec<_>>();
+
+        let catalog = TemplateCatalog { standard_templates, custom_templates };
+        Ok(catalog)
```
Comment (cmyr) [inline src/instruments.rs:130]: okay, after some testing I have some concerns about this method:

- the output of `xcrun xctrace list templates` and `instruments -s templates` differs. The former looks like what you're checking for, but the latter looks like:

```
$ instruments -s templates

Known Templates:
"Activity Monitor"
"Allocations"
"Animation Hitches"
"App Launch"
"Blank"
"Core Data"
"Counters"
"Energy Log"
"File Activity"
"Game Performance"
"Leaks"
"Logging"
"Metal System Trace"
"Network"
"SceneKit"
"SwiftUI"
"System Trace"
"Time Profiler"
"Zombies"
"~/Library/Application Support/Instruments/Templates/MyTemplate.tracetemplate"
```

In additon, on my laptop (macOS `10.15.7`, Xcode `12.4`) The output of `xcrun xtrace list templates` is printed to *stderr*, not *stdout*. I guess this changed at some point?

In any case, it looks like these tools will, at the very least, require separate parsing code; and in the case of `xcrun xctrace` if stdout is empty but stderr is not, I would use stderr instead of stdout.

Additionally, I'm curious if there is any practical different between the standard and custom templates? Do we treat them differently in any way?

File: src/instruments.rs [hunk 16]
```diff
@@ -136,12 +317,69 @@ fn resolve_template(args: &Opts) -> &str {
...
+    xctrace_tool: &XcodeInstruments,
+    app_config: &AppConfig,
+    workspace: &Workspace,
+) -> Result<PathBuf> {
+    // 1. Get the template name either from config or interactive menu
+    let template_name: String = match app_config.template_name {
+        Some(ref name) => resolve_template_name(name).to_owned(),
+        None => {
+            let template_catalog = xctrace_tool.available_templates()?;
+            select_template(&template_catalog).unwrap()
```
Comment (cmyr) [inline src/instruments.rs:344]: So having played around with this, I have mixed feelings.

In my profiling experience, 90% of the time you probably want the `time` profiler, and then sometimes alloc/leaks, and then everything else only in very particular circumstances. I think that throwing the user into the full list of options, without guidance, is not a great user experience.

Although I do like the polished feel, I think my *personal* inclination would be to default to `time` if no argument is provided. If we really do want to show the chooser, I would make sure `Time Profiler` was selected by default, and I would try to do something to indicate that it was the recommended template.

Basically my higher-level concern is that showing the picker does not actually improve the user experience? I'm happy to have a conversation about this, though!

**Review by cmyr — COMMENTED** [review]
Okay, a couple more notes after getting this running on my machine. :)

File: src/instruments.rs (hunk 15 above)
Comment (bosr) [reply src/instruments.rs:130]: For the output of `/usr/bin/instruments`: Argh, absolutely true, sorry about that. I'll soon add the missing parsing logic.

Yes, the output of `xcrun xctrace` changed to stdout, at some point (macOS 11.0?) I saw it on a forum, but I forgot to change that, sorry again, and thanks for your throughout testing. I guess that's an easy fix. I'm happy to apply the fix you described, but we might have to iterate as I have no way to test it on 10.15 anymore.

Yes you're right, the 2 kinds of templates are treated identically. The separation is just for bookeeping, but that's all.

I went for supporting custom templates for completeness, and also because [custom instruments](https://developer.apple.com/wwdc18/410) are pretty cool.

**Comment by bosr** [comment]
Sure! Thank you for taking the time to work with me! After checking the latest behavior, just tell me and I'll remove the menu in no time.

**Comment by cmyr** [comment]
Okay, this is working for me.

I think the improvements to the menu make it much more useful, and I don't mind keeping it, although I do think a full 'interactive mode' could be fun?

**Comment by bosr** [comment]
Hello @cmyr, true, with the improvements, the menu is more integrated in the workflow and feels coherent with the default value. Thanks!

Regarding the interactive mode, I see two approaches: 
- if a `--interactive` flag is provided, have the user go through selection menus, and then perform the profiling
- introduce a config file which the user can fill interactively just once:
  - the config could be `trac(e|ing).toml` or `instruments.toml`
  - or it could be a `[package.metadata]` section in `Cargo.toml`: I tried to do this but I failed at accessing the data programmatically - you can certainly do better than me here 😅

I'll now squash my commits for you to merge, thanks again! 🎉

**Comment by cmyr** [comment]
I think interactive should show the menus each time, if you know what you want it should be easy enough to just pass the flags on the CLI?

**Comment by bosr** [comment]
Makes sense, and that's also the simplest to implement.

**Comment by cmyr** [comment]
And I think that with adding interactive mode, the big question becomes 'what happens if you pass no arguments'?

Currently, this just runs the time profiler on the main executable. That might also not be a great experience? I'm not sure. It might be easiest if we always just show the help message and exit when no arguments are provided? One concern with defaulting to interactive mode is that it makes discovering the flags more difficult.

**Comment by bosr** [comment]
Yes, I really like that. In that perspective, do you want me to remove the interactive menu and the default template?

**Comment by cmyr** [comment]
yea, I think that makes sense, then we can do the big UI overhaul all at once. :)

**Comment by bosr** [comment]
Hello! Here is the update we agreed upon, and the behavior in the 3 cases

```
$ cargo instruments -l
Xcode Instruments templates:

built-in            abbrev
--------------------------
Activity Monitor
Allocations         (alloc)
Animation Hitches
App Launch
Core Data
Counters
Energy Log
File Activity       (io)
Game Performance
Leaks
Logging
Metal System Trace
Network
SceneKit
SwiftUI
System Trace        (sys)
Time Profiler       (time)
Zombies

custom
--------------------------
MyTemplate

$ cargo instruments
Missing template name

$ cargo instruments -t time
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
   Profiling target/debug/tries with template 'Time Profiler'
  Trace file target/instruments/tries_Time-Profiler_2021-05-15T19:16:51.trace
```

Ping me you're good with this, I'll push the code right away.

**Comment by cmyr** [comment]
The only thing to think about is the "no arguments" case.

I would suggest, in this case, that we print the full `--help` message, but it might also be useful to add an additional comment, saying something like "to use the Time Profiler, use 'cargo instruments -t time'"?

**Comment by bosr** [comment]
Yes, I tried that, but I don't know how I can get clap print the App's help message. I know the [write_help](https://docs.rs/clap/2.33.3/clap/struct.App.html#method.write_help) method on clap `App`, but at this point in the code, if I'm not mistaken, we cannot access the clap App because it is completely abstracted away by `structopt`. And that's a good thing!

Let me summarize the problem:
> if we want the help message to appear when the template name is missing, then the `template_name` should not be an `Option<String>`, but simply a `String`. However, if `template_name` is a string, then the `cargo instruments --list-templates` is invalid input (missing template_name).

If `list` was a subcommand, and `profile` another subcommand, there would be no option or conflict, and the help message would be displayed without any additional app logic (you know that, obviously, I'm just summarizing).

In the next iteration, we could have this CLI api:
- list the templates with `cargo instruments list`
- get an error message with `cargo instruments`
- alternatively
  - get an error message with `cargo instruments profile`
  - or profile using the default template of `time` with `cargo instruments profile`
- profile with `cargo instruments profile alloc`
- interactive menu with `cargo instruments profile --interactive`

I can even implement this right now, leaving the implementation of `--interactive` for a next PR.

What do you think?

**Comment by cmyr** [comment]
hmm, I think there should be *some* way to make this work... i'll poke around for a sec

**Comment by cmyr** [comment]
Okay, I think you can get this working by adding the `required_unless = "list_templates"` attribute to the `structopt` annotation on the `template_name` field in `AppConfig`.

This doesn't seem to print the full help; not sure if there's another option for that?

**Comment by bosr** [comment]
Thanks, `required_unless` made it work.

**Review by cmyr — APPROVED** [review]
Okay two tiny things and then I'm ready to merge!

File: src/opt.rs [hunk 17]
```diff
@@ -11,58 +11,76 @@ pub(crate) enum Cli {
     ///
     /// By default, cargo-instruments will build your main binary.
     #[structopt(name = "instruments")]
-    Instruments(Opts),
+    Instruments(AppConfig),
```
[the remark this answers is not in the document]
Comment (bosr) [reply src/opt.rs:14]: Thanks!

**Comment by bosr** [comment]
All good, thanks!

**Comment by bosr** [comment]
I fixed the failing tests.

**Comment by cmyr** [comment]
Great, thanks!

**Comment by bosr** [comment]
@cmyr, thanks!!

Would you consider publishing a new tag and homebrew formula for this?

**Comment by cmyr** [comment]
Definitely. Are you interested in trying to do interactive mode in the near-term? I suppose in any case it's probably worth publishing now, interactive mode won't be a breaking change.

Thanks for all of your work on this!

**Comment by bosr** [comment]
Hello! Thanks, but I don't think I can in the near-term. Please go ahead!
